### PR TITLE
doctrine:migration:diff does not work if the table has an index with a functional part.

### PR DESCRIPTION
**Pull request by prohalexey**
|      Q       |   A
|------------- | -----------
| Type         | bug

#### Summary

if the table has an index with a functional part, command doctrine:migration:diff fails with error 

```
In Index.php line 81:
                                                                               
  Doctrine\DBAL\Schema\Index::_addColumn(): Argument #1 ($column) must be of   
  type string, null given, called in /opt/core/vendor/doctrine/dbal/src/Schem  
  a/Index.php on line 72 
```

That's because column name for the functional part is null.

### CODE REVIEW

**Comment by derrabus** [comment]
Thank you. Please note that we don't merge any changes without tests.

**Comment by prohalexey** [comment]
@derrabus What tests do you think will be suitable for this patch?

**Comment by derrabus** [comment]
A test that reproduces your issue and fails without your change.

File: src/Schema/MySQLSchemaManager.php [hunk 1]
```diff
@@ -619,4 +622,21 @@ private function parseCreateOptions(?string $string): array
...
+     * For a functional key part, COLUMN_NAME column is NULL and EXPRESSION indicates the expression for the key part.
+     */
+    private function getColumnNameForIndexFetch(): string
+    {
+        if ($this->_platform instanceof MySQL8013Platform) {
+            return "COALESCE(COLUMN_NAME, concat('(', EXPRESSION, ')')) AS Column_Name";
+        }
+
+        return 'COLUMN_NAME as Column_Name';
+    }
```
Comment (derrabus) [inline src/Schema/MySQLSchemaManager.php]: Please move this function to the platform instead. We don't want to clutter the codebase with these "if platform version is x then this else that" blocks and that's exactly the problem that the platform classes solve for us.

File: src/Platforms/AbstractPlatform.php [hunk 2]
```diff
@@ -396,6 +396,11 @@ abstract public function getBlobTypeDeclarationSQL(array $column);
      */
     abstract public function getName();
 
+    /**
+     * A flag that indicates whether the platform supports functional indexes.
+     */
+    abstract public function isSupportsFunctionalIndex(): bool;
```
Comment (derrabus) [inline src/Platforms/AbstractPlatform.php]: 1. `isSupportsFunctionalIndex()` is an odd name. `supportsFunctionalIndex()` maybe?
2. Adding a new abstract method to the platform is a BC break for packages that add their own platforms. We should avoid this. Can we come up with a default implementation instead?
3. Does this method have to be public? As far as I can tell, it's only ever called by the platform itself.
4. Do we really need this method?

File: src/Platforms/MySQL8013Platform.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Doctrine\DBAL\Platforms;
+
+/**
+ * Provides features of the MySQL since 8.0.13 database platform.
+ *
+ * Note: Should not be used with versions prior to 8.0.13.
+ *
+ * This class will be merged with {@see MySQLPlatform} when support for MySQL
+ * releases prior to 8.0.13 will be dropped.
```
Comment (derrabus) [inline src/Platforms/MySQL8013Platform.php]: Please remove this remark. The class is not deprecated yet.

File: src/Schema/AbstractSchemaManager.php [hunk 4]
```diff
@@ -1470,7 +1471,7 @@ protected function _getPortableTableIndexesList($tableIndexes, $tableName = null
                 ];
             }
 
-            $result[$keyName]['columns'][]            = $tableIndex['column_name'];
+            $result[$keyName]['columns'][]            = stripslashes($tableIndex['column_name']);
```
Comment (derrabus) [inline src/Schema/AbstractSchemaManager.php]: Why is this change necessary?

File: src/Schema/Index.php [hunk 5]
```diff
@@ -334,6 +351,11 @@ public function getOptions()
         return $this->options;
     }
 
+    public static function isFunctionalIndex(string $name): bool
+    {
+        return $name[0] === '(' && $name[mb_strlen($name) - 1] === ')';
```
Comment (derrabus) [inline src/Schema/Index.php]: Please use `substr()` to fetch the last character of `$name`.

**Review by derrabus — CHANGES_REQUESTED** [review]
Please add functional tests. The tests you've added never hit a real database and thus don't tell us that your feature actually works.

File: src/Driver/AbstractMySQLDriver.php [hunk 6]
```diff
@@ -69,7 +70,11 @@ public function createDatabasePlatformForVersion($version)
                     );
                 }
 
-                return new MySQL80Platform();
+                if (version_compare($version, '8.0.13', '<')) {
+                    return new MySQL80Platform();
+                }
```
Comment (derrabus) [inline src/Driver/AbstractMySQLDriver.php:75]: We have tests that make sure we pick the right platform for a given version string. Please update that test for your new platform class.

File: src/Schema/MySQLSchemaManager.php (hunk 1 above)
Comment (prohalexey) [reply src/Schema/MySQLSchemaManager.php]: Done

File: src/Platforms/AbstractPlatform.php (hunk 2 above)
Comment (prohalexey) [reply src/Platforms/AbstractPlatform.php]: 1. Done
2. Done
3. Done
4. I think yes, because various databases already support this feature and others may launch it in new versions.

File: src/Platforms/MySQL8013Platform.php (hunk 3 above)
Comment (prohalexey) [reply src/Platforms/MySQL8013Platform.php]: Done

File: src/Schema/AbstractSchemaManager.php (hunk 4 above)
Comment (prohalexey) [reply src/Schema/AbstractSchemaManager.php]: Example: 
The expression in functional part of index could be something like
```
(cast(json_extract(`path`,_utf8mb4'$[*]') as char(255) array))
```
So, then you get info from statistic
```
SELECT
     COALESCE(COLUMN_NAME, concat('(', EXPRESSION, ')')) AS Column_Name
FROM information_schema.STATISTICS
```
You've got 
```
(cast(json_extract(`path`,_utf8mb4\'$[*]\') as char(255) array))
```
Quotes will be backslashed.

Comment (derrabus) [reply src/Schema/AbstractSchemaManager.php]: That query of yours is specific to MySQL, but you run the `stripslashes()` on the abstract schema manager. That feels odd.

File: src/Schema/Index.php (hunk 5 above)
Comment (prohalexey) [reply src/Schema/Index.php]: Theoretically, the expression can contain non-Latin characters. For example, this is valid statement
```
CREATE FUNCTION АБ⍅ (s CHAR(20))
    RETURNS CHAR(50) DETERMINISTIC
	RETURN CONCAT('Hello, ',s,'!');
```

File: src/Schema/AbstractSchemaManager.php (hunk 4 above)
Comment (prohalexey) [reply src/Schema/AbstractSchemaManager.php]: Ok, maybe I could remove backslashes in the SQL Query.
Done

File: src/Schema/Index.php (hunk 5 above)
Comment (prohalexey) [reply src/Schema/Index.php]: Replaced with `mb_substr`

File: src/Driver/AbstractMySQLDriver.php (hunk 6 above)
Comment (prohalexey) [reply src/Driver/AbstractMySQLDriver.php:75]: Ok. Done

File: src/Schema/Index.php (hunk 5 above)
Comment (derrabus) [reply src/Schema/Index.php]: Yes, but `)` iself is an ASCII character. `substr()` is safe here. We don't want to depend on ext-mbstring to be present.

Comment (prohalexey) [reply src/Schema/Index.php]: Done

**Comment by prohalexey** [comment]
@derrabus Could you give some advice or explanations about falling tests? I tried running tests on my local machine against postgresql and mysql and they passed. As I see, there is a lot of errors, and i don't see the correlation between the errors and my code. What could be a root cause of problem ?
